### PR TITLE
New search engine: wordnik.com

### LINKED ISSUE #1844: Feature suggestion: Add Wordnik
Hi there, would it be possible to add Wordnik? This is where DuckDuckGo gets its definitions from, and they're much better than Wikipedia/Wiktionary. It would be nice to have this for instant answers on the sidebar, with top priority above other services that offer definitions, such as Wikipedia and Wolfram|Alpha. Thanks!

### PR DESCRIPTION
**Pull request by plague-doctor**
## What does this PR do?

Adds a new search engine Wordnik.com to retrieve word definitions. Similar to DuckDuckGo query `define: word`.

## Why is this change important?

Current version of searx missing the functionality known from DuckDuckGo. This PR eliminates this gap.

## How to test this PR locally?

Run `make run`.

## Related issues

Closes #1844

### CODE REVIEW

File: searx/engines/wordnik.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        return[]
+
+    results = []
+    word_defs = []
+    word_defs_ib = []
+    urls = []
+    lines = ""
+    lines_ib = "<div>"
+
+    word_xpath = '//*[@id="headword"]/text()'
```
Comment (kvch) [inline searx/engines/wordnik.py]: These four variables can be moved to the global level of this file.

File: searx/engines/wordnik.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+def response(resp):
+    if resp.status_code == 404:
+        return []
+
+    try:
+        dom = fromstring(resp.text)
+    except Exception:
+        raise_for_httperror(resp)
+        return[]
```
Comment (kvch) [inline searx/engines/wordnik.py]: Tyop:
```python
return []
```

File: searx/engines/wordnik.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    for definition in definitions:
+        for i in range(len(definition)):
+            definition_extra_ib = ""
+            definition_extra = ""
+
+            part_of_speach_xpath = f"./li[{i+1}]/abbr"
+            definition_extra_xpath = f"./li[{i+1}]/i"
+            definition_text_xpath = f"./li[{i+1}]"
+
+            part_of_speach = extract_text(definition.xpath(part_of_speach_xpath))
```
Comment (kvch) [inline searx/engines/wordnik.py]: Nit: isn't `speach` supposed to be `speech`?

File: searx/engines/wordnik.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+SEARCH_URL = URL + '/words/{query}'
+
+
+def request(query, params):
+    params['url'] = SEARCH_URL.format(query=query)
+    return params
+
+
+def response(resp):
+    if resp.status_code == 404:
```
Comment (kvch) [inline searx/engines/wordnik.py]: I think HTTP 404 is handled by the online plugin, there is no need for this check.

**Comment by kvch** [comment]
Awesome PR!

File: searx/engines/wordnik.py (hunk 1 above)
Comment (plague-doctor) [reply searx/engines/wordnik.py]: Fixed.

File: searx/engines/wordnik.py (hunk 4 above)
Comment (plague-doctor) [reply searx/engines/wordnik.py]: OK.

File: searx/engines/wordnik.py (hunk 3 above)
Comment (plague-doctor) [reply searx/engines/wordnik.py]: Now you know why do I need wordnik :-)

File: searx/engines/wordnik.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    results = []
+    word_defs = []
+    word_defs_ib = []
+    urls = []
+    lines = ""
+    lines_ib = "<div>"
+
+    word = extract_text(dom.xpath(word_xpath))
+    definition_from = extract_text(dom.xpath(from_xpath))
```
Comment (return42) [inline searx/engines/wordnik.py]: There are multiple definition_from ...

- from The American Heritage® Dictionary of the English Language, 5th Edition.
- from The Century Dictionary.
- ...

File: searx/engines/wordnik.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    lines_ib += f"<small>{definition_from}</small></div>"
+
+    urls.append({'title': word, 'url': f"{URL}/words/{word}"})
+
+    results.append({
+        'infobox': word,
+        'content': lines_ib,
+        'urls': urls})
+
+    results.append({
```
Comment (return42) [inline searx/engines/wordnik.py:72]: IMO this is not needed / _infobox_ should be enough I think.

File: searx/engines/wordnik.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    "results": 'HTML',
+}
+
+categories = ['general']
+paging = False
+
+word_xpath = '//*[@id="headword"]/text()'
+from_xpath = '//*[@id="define"]/div/h3[1]'
+definitions_xpath = '//*[@id="define"]/div/ul[1]'
+misspelling_xpath = '//*[@id="define"]/div/ul/li/xref'
```
Comment (return42) [inline searx/engines/wordnik.py]: Do you have an example (search term) matching this xpath? .. I tried several typos, but I did not find a term where I got a misspelling_xpath in the response from wrodnick.

File: searx/engines/wordnik.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+def request(query, params):
+    params['url'] = SEARCH_URL.format(query=query)
+    return params
+
+
+def response(resp):
+    try:
+        dom = fromstring(resp.text)
+    except Exception:
```
Comment (return42) [inline searx/engines/wordnik.py]: Never catch general exception.

File: searx/engines/wordnik.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        raise_for_httperror(resp)
```
Comment (return42) [inline searx/engines/wordnik.py]: This is only for HTTP errors (see my branch)

File: searx/settings.yml [hunk 10]
```diff
@@ -1264,6 +1264,14 @@ engines:
     categories: videos
     disabled : True
 
+  - name: wordnik
+    engine: wordnik
+    shortcut: def
+    base_url: https://www.wordnik.com/
+    categories: general
+    timeout: 5.0
+    disabled: False
```
Comment (return42) [inline searx/settings.yml]: I vote for disabling by default, since wordnik is english only and not all of our users are looking for english terms.

**Review by return42 — CHANGES_REQUESTED** [review]
Thanks for this engine, I really like it but I think there are some quirks we should avoid, see my comments.  

In my branch https://github.com/return42/searx/commits/add-wordnik I rebased and squash your commits.

On top of this branch you find a commit that simplifies the implementation and fixes the quirks.

If it is OK for you merge my commit in this PR.

File: searx/engines/wordnik.py (hunk 5 above)
Comment (plague-doctor) [reply searx/engines/wordnik.py]: Sometimes the definitions take huge amount of space. I just wanted to mimic DuckDuckGo behavior and return only a few definitions. Just to get general meaning of the word not the whole etymology or every possible idiom. There was a link to the wordnik for. But I am quite happy with your approach.

File: searx/engines/wordnik.py (hunk 7 above)
Comment (plague-doctor) [reply searx/engines/wordnik.py]: For example `distopia`. It is easier for us - ESLs :-)

File: searx/settings.yml (hunk 10 above)
Comment (plague-doctor) [reply searx/settings.yml]: Absolutely! Left over from testing. Sorry.

File: searx/engines/wordnik.py (hunk 6 above)
Comment (plague-doctor) [reply searx/engines/wordnik.py:72]: Agree. Your take is much more simple and neat :-)

File: searx/engines/wordnik.py (hunk 8 above)
Comment (plague-doctor) [reply searx/engines/wordnik.py]: Yep.

File: searx/engines/wordnik.py (hunk 9 above)
Comment (plague-doctor) [reply searx/engines/wordnik.py]: Lots to be learnt. Will get there eventually.

File: searx/engines/wordnik.py (hunk 7 above)
Comment (return42) [reply searx/engines/wordnik.py]: Ah, Ok .. thanks .. I think it is OK to display it "as is" ..

image

**Comment by plague-doctor** [comment]
Opss... I have screwed this up... Will create a new PR. Sorry.

**Pull request closed without merging**